### PR TITLE
Implement Group 3 Test: kNN method

### PR DESCRIPTION
**Pull request by mbobra**
**Description**
This PR implements an outlier detection test using a kNN method described in [Talagala+2019](http://dx.doi.org/10.1029/2019WR024906). Group 3 of the [Outlier Detection Test Best Practices](https://docs.google.com/document/d/1YmXWDMHCKnWniQ4sMkkT5g5j5H5TmT3f/edit) document suggests using machine learning methods for outlier detection tests. Right now, we are using univariate data (pH). Ideally, we could detect outliers using multivariate methods. 

We do not have any labelled data. We should ask Daniel Wisheropp for true and false outliers for historical data. Right now, we are reading data that was artificially flagged using the outlier detection tool. Here are the data we use for this model, which shows 5857 inliers and 444 outliers: 

initialdataset

Next, we create simple features in the temporal domain using the Python package [tsfresh](https://tsfresh.readthedocs.io/en/latest/). See [here](https://tsfresh.readthedocs.io/en/latest/text/list_of_features.html) for a list of available features. We will begin with a subset of six features. In the future, we can look at the spectral domain (e.g. Fourier space). Future ideas also include computing features with convolutional kernels (e.g. [ROCKET](https://www.aeon-toolkit.org/en/stable/examples/transformations/rocket.html)). 

We split the data by time: All the data prior to January 23, 2016 goes into the training set, and the rest of the data go into the test set. Here are some initial model results:
* False positives: 9
* False negatives: 132
* True positives: 86
* True negatives: 2474

The following plot shows a decision boundary for a two-dimensional feature space. One feature describes the autocorrelation of a week-long signal with the prior week; the other feature is simply the pH value itself. For all possible slides through the seven-dimensional feature space, see the Jupyter notebook.

kNN_decision_boundary

**Tasks**
- [x] Create synthetic labelled data (`flagged_sample_data_BKS.csv`)
- [x] Derive simple features from univariate time series data
- [x] Split the time series into training and testing data (Question: Can we use time series from multiple stations to train the model, or does every station show idiosyncratic behavior?)
- [x] Train kNN model using simple features (in other words, lose the time dependence and identify outliers in feature space)
- [x] Test kNN model
- [x] Validate kNN model

### CODE REVIEW

**Comment by mbobra** [comment]
@JamesSLogan I'll add you for a review -- for this code, you can feel free to do more of a sanity check than a careful assessment. If we plan to include this in the tool, I'll port it over to `od.py` (and write unit tests) and we should check that code carefully.

File: dwr/dwr.ipynb
Comment (JamesSLogan) [inline dwr/dwr.ipynb:1]: I think this cell can be removed

**Review by JamesSLogan — APPROVED** [review]
Looks great!

File: dwr/dwr.ipynb
Comment (mbobra) [reply dwr/dwr.ipynb:1]: Nice catch, thanks.
